C API

kripken opened the pull request:
This adds a C API and a test which shows an example of using it. I've never been much of an API designer, apologies for what you're about to see.

The API is generally similar to LLVM's API which is probably the most popular compiler API. However, as we have structured control flow and statements=expressions etc., this API follows that. As a follow-up, though, I intend to add a basic-block oriented API on top of this, which should be more familiar for compilers (and then using relooping internally we can generate structured control flow as needed).

The "hello world" example and its output are the last two files in the diff.

binji commented on src/binaryen-c.h:
```diff
@@ -0,0 +1,246 @@
+
+typedef uint32_t BinaryenType;
+
+BinaryenType BinaryenNone(void);
```
Seems to not just use an enumeration here. Is this to prevent API breakage if the type values change?

kripken replied:
Also for changes, but the main reason is that compilers using the C API might not want to use a C *header*. For example, Rust is not written in C, so it can't execute the C header to get those enum values. Instead it links with C libraries and calls methods on them using the C ABI.

sunfishcode replied:
Another operation is a C++11 enum class with an explicit type:

```
enum class Color : uint32_t
{
    Red, Green, Blue
};
```

kripken replied:
That still doesn't help the Rust use case, though? Or did I misunderstand you?

sunfishcode replied:
It might make the C++ use case a little nicer while not harming the rust use case.

kripken replied:
I see. Yeah, I agree that might be nicer. There is the gcc bug mentioned in another comment on this PR, though, which worries me.

binji commented on src/binaryen-c.h:
```diff
@@ -0,0 +1,246 @@
+
+typedef int32_t BinaryenOp;
+
+BinaryenOp BinaryenClz(void);
```
Same here, this would be nicer as an enum.

binji commented on src/binaryen-c.h:
```diff
@@ -0,0 +1,246 @@
+
+typedef void* BinaryenFunctionTypeRef;
+
+BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name, BinaryenType result, BinaryenType* paramTypes, BinaryenIndex numParams);
```
You could use varargs for this for a slightly friendlier API

kripken replied:
Yeah, I can see an argument for varargs. But reasons against are that the LLVM API doesn't use it, and also there are probably ABI corner cases (maximum elements, maybe?).

binji replied:
You could have both with and without varargs. But it's probably OK to just have this.

binji commented on src/binaryen-c.h:
```diff
@@ -0,0 +1,246 @@
+typedef void* BinaryenExpressionRef;
+
+// Block: name can be NULL
+BinaryenExpressionRef BinaryenBlock(BinaryenModuleRef module, const char* name, BinaryenExpressionRef* children, BinaryenIndex numChildren);
```
Probably nicer to have a way to add child expressions incrementally. Otherwise the API user will have to create their own array externally, only to have it allocate additional memory internally.

kripken replied:
Interesting point. I think you're right and we should probably have both. Although then it seems we'd also want an API to set the size of the basic block before incrementally adding to it (otherwise the vector contents would get reallocated)?

binji replied:
Yeah, you could have a size hint parameter in the constructor.

kripken replied:
Sounds good. First we need to add support for reserving space in arena-allocated arrays, then we could do this.

binji commented on src/binaryen-c.h:
```diff
@@ -0,0 +1,246 @@
+// Literals. These are passed by value.
+
+struct BinaryenLiteral {
+  int type; // size of enum in c++
```
Might want to make an enum class so you can specify its size.

kripken replied:
Hmm, yeah, we should do that. Also it might save space since it fits in a byte.

However I think this might need to wait, as the gcc in Ubuntu 14.04 LTS has a bug on enum classes with int8_t that I've seen hit (e.g. in SpiderMonkey). Unless we do int8_t for the storage (no enum class, just plain int8_t) and an enum for the values, and do some casting, which maybe isn't so bad since it's only done internally in wasm.h?

binji commented:
Looks pretty good. Have you thought about how you'd manipulate the AST after it has been created? Also, personally I'd suggest something shorter than `Binaryen` for the prefix. But maybe that's just me. :)

kripken commented:
No, I haven't thought much about manipulation yet. Hopefully most compilers won't need it? Probably that's too optimistic ;)

About the prefix, how about "Bin"? Or any other ideas?

binji commented:
> No, I haven't thought much about manipulation yet. Hopefully most compilers won't need it? Probably that's too optimistic ;)

Well, if you imagine the users of the API are just producers than yeah, probably not needed. I was imagining users who will want to read a binary module and then do stuff with it.

> About the prefix, how about "Bin"? Or any other ideas?

How about `BYE` :)

jfbastien commented:
> > About the prefix, how about "Bin"? Or any other ideas?

> How about `BYE` :)

`Known` would be nice: "the prefix? It is `Known`."

kripken commented:
> How about BYE :)

`BYE` suggests `BYOB` to me... so, Bring Your own... Expression? ;)

But I like the shortening, `BYE` for BinarYEn. Another option might be `BYN`?

Probably too short, but maybe, `B_`?

> Known would be nice: "the prefix? It is Known."

Definitely the kind of pun I like, but probably too much for a C API ;)

kripken commented:
Or maybe `BIN`?

binji commented:
I like `BYN` better. `BIN` looks like it could be anything, but I can't think of anything else that looks like `BYN`.

kripken commented:
Yeah, maybe the choice is `BYN` (or `Byn`?) vs `Binaryen`. I'm ok with either. If no one chimes up I suppose we can leave it as `Binaryen` for now.

Meanwhile I added a kitchen-sink test of the api, which found a few bugs that are now fixed.

kripken commented:
Ok, let's merge this. We can change the name later if there are opinions about it.
